src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/events/ActionPlanEventPublisher.kt

```diff
@@ -0,0 +1,38 @@
+package uk.gov.justice.digital.hmpps.hmppsinterventionsservice.events
+
+import org.springframework.context.ApplicationEvent
+import org.springframework.context.ApplicationEventPublisher
+import org.springframework.stereotype.Component
+import org.springframework.web.bind.annotation.GetMapping
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.component.LocationMapper
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.controller.ActionPlanController
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.entity.ActionPlan
+import kotlin.reflect.KFunction
+
+enum class ActionPlanEventType {
+  SUBMITTED,
+}
+
+class ActionPlanEvent(source: Any, val type: ActionPlanEventType, val actionPlan: ActionPlan, val detailUrl: String) : ApplicationEvent(source) {
+  override fun toString(): String {
+    return "ActionPlanEvent(type=$type, actionPlanId=${actionPlan.id}, detailUrl='$detailUrl', source=$source)"
+  }
+}
+
+@Component
+class ActionPlanEventPublisher(
+  private val applicationEventPublisher: ApplicationEventPublisher,
+  private val locationMapper: LocationMapper
+) {
+
+  fun actionPlanSubmitEvent(actionPlan: ActionPlan) {
+    applicationEventPublisher.publishEvent(ActionPlanEvent(this, ActionPlanEventType.SUBMITTED, actionPlan, createDetailUrl(actionPlan)))
```

Comment:
(Sorry if this has been discussed already.)

What do you think about using `publishEvent(ActionPlanSubmitted(this, actionPlan))`? As in, use distinct event classes in favour of event type?

That way all submission-specific concerns (like URLs) can be encapsulated in the event itself.

Reply:
I do like that idea a lot!!!! ... would prefer to raise it as a separate tech debt item as the amount of change already in this PR is significant. Would that be ok?

Reply:
of course, this isn't a blocker at all

Reply:
Thanks and I wish I had done it in the first place. Will stick a comment in the code too.